### DISCUSSION TITLE
send a zip file as an attachment in a response

### DISCUSSION THREAD

**Discussion by hrtejada** [fastapi/fastapi#8026] [Questions] [answered]
Is it possible to send a zip file as an attachment in a response? Looked through the starlette's response library but could not figure out a way to do this. My initial feeble attempt at making this work was using the Response object, specificying the media_type as 'application/zip' and passing the zipped I/O buffer in the content...normally in Flask I would do something like:

```
response_obj = Response(zipped, mimetype='application/zip', headers={
            'Content-Disposition': 'attachment;filename=api.zip'
            })
return response_obj
```

Where zipped was of type io.BytesIO(). Is something like this possible in FastAPI or using Starlette's responses?

**Comment by dmontagu** [comment]
It should be possible using Starlette's responses, but I'm not sure how high-level of an API there is for this. Either way, I think this would almost certainly be making direct (or nearly direct) use of starlette, rather than FastAPI, so you might have more luck asking there.

**Comment by euri10** [comment, marked as answer]
as I was commenting on https://github.com/encode/starlette/issues/609 it does indeed work with a FileResponse
```
@router.get("/zip")
def zip():
    content = b"<file content>" * 1000
    with open("/tmp/stuff", "wb") as tozip:
        tozip.write(content)
    with ZipFile('example.zip', 'w') as myzip:
        myzip.write("/tmp/stuff")
    response = FileResponse(path='example.zip', filename="example.zip")
    return response
```

**Comment by euri10** [comment]
works in swagger too for more swag
![img](https://i.imgur.com/JJ2dJIV.png)

**Comment by hrtejada** [comment]
Worked like a charm! 👌

**Comment by tiangolo** [comment]
Thanks @euri10 and @dmontagu ! :clap: :bowing_man: 

Thanks @hrtejada for reporting back and closing the issue.

**Comment by Yassunaga** [comment]
Is there a way to return the zip file from memory, without writing it to disk?

**Comment by rule88** [comment]
Would like to know this as well indeed. I am trying to return a BytesIO object (zip) as a file to download directly, but instead it is return as inline?

![image](https://user-images.githubusercontent.com/11073249/105477664-7312c780-5ca2-11eb-8a47-bc2d53e01c72.png)

```python
    @app.post("/zip")
    def zip():
    ...
    zipfile = utils.zip_dir(zipfile_name, result_dir)
    zipfile.seek(0)
    return StreamingResponse(zipfile, media_type="application/x-zip-compressed",
    headers={'Content-Disposition': f'attachment; filename="{zipfile_name}"'})
```

**Reply by shenwpo** [reply]
```
def download(folder_path: str):
    output = BytesIO()
    with ZipFile(output, "w", ZIP_DEFLATED) as zip_file:
        for file in Path(folder_path).glob("**/*"):
            zip_file.write(file, file.relative_to(folder_path))

    download_name = folder_path.replace("/", "_")
    name = urllib.parse.quote(download_name)
    return Response(
        output.getvalue(),
        media_type="application/zip",
        headers={"Content-Disposition": f"attachment; filename={name}.zip"},
    )
```

**Comment by Kludex** [comment]
This issue was solved. If you have different questions, create a new one, please.
